pymc_marketing/customer_choice/nested_logit.py

```diff
@@ -0,0 +1,783 @@
+                        f"prod_{n}_m", (P_nest[:, pt.newaxis] * P_y_given_nest)
+                    )
+                    if parent in path_prods_m:
+                        path_prods_m[parent].append(prod)
+                    else:
+                        path_prods_m[parent] = []
+                        path_prods_m[parent].append(prod)
+                    if is_last:
+                        P_ = pm.Deterministic(
+                            f"P_{parent}_children",
+                            pm.math.concatenate(path_prods_m[parent], axis=1),
+                        )
+                        child_nests[parent] = P_
+            else:
+                child_nests = {}
+
+            ## Top Level
+            cond_prob_t, nest_prob_t = self.make_P_nest(
+                U, W, betas_fixed_, lambdas_nests, "top"
+            )
+
+            path_prods_t = []
+            ordered = [(key, min(vals)) for key, vals in nest_indices["top"].items()]
+            top_nests = [x[0] for x in sorted(ordered, key=lambda x: x[1])]
+            for _idx, n in enumerate(top_nests):
+                P_nest = nest_prob_t[n]
+                P_y_given_nest = cond_prob_t[n]["P_y_given"]
+                if n in child_nests:
+                    P_y_given_nest_mid = pm.Deterministic(
+                        f"P_y_given_nest_mid_{n}", child_nests[n]
+                    )
+                    prod = pm.Deterministic(
+                        f"prod_{n}_t", (P_nest[:, pt.newaxis] * (P_y_given_nest_mid))
+                    )
+                else:
+                    prod = pm.Deterministic(
+                        f"prod_{n}_t", (P_nest[:, pt.newaxis] * (P_y_given_nest))
+                    )
+                path_prods_t.append(prod)
+            P_ = pm.Deterministic(
+                "p", pm.math.concatenate(path_prods_t, axis=1), dims=("obs", "alts")
+            )
+
+            _ = pm.Categorical("likelihood", p=P_, observed=y_data, dims="obs")
+
+        self.model = model
+        return model
+
+    def _data_setter(
+        self,
+        X: np.ndarray | pd.DataFrame,
+        y: np.ndarray | pd.Series | None = None,
+    ) -> None:
+        """Set the data.
+
+        Required from the parent class
+
+        """
+
+    def create_idata_attrs(self) -> dict[str, str]:
+        """Create the attributes for the InferenceData object.
+
+        Returns
+        -------
+        dict[str, str]
+            The attributes for the InferenceData object.
+
+        """
+        attrs = super().create_idata_attrs()
+        attrs["covariates"] = json.dumps(self.covariates)
+        attrs["depvar"] = json.dumps(self.depvar)
+        attrs["choice_df"] = json.dumps("Placeholder for DF")
+        attrs["nesting_structure"] = json.dumps(self.nesting_structure)
+        attrs["utility_equations"] = json.dumps(self.utility_equations)
+
+        return attrs
+
+    def sample_prior_predictive(self, extend_idata, kwargs):
```

Comment:
type hints missing

Reply:
added but i had to override parent class function signature because mypy was complaining